pageserver/src/tenant/timeline.rs

```diff
@@ -1288,88 +1301,95 @@ impl Timeline {
-            Err(e) => {
-                // Non-fatal.  Shutdown is infallible.  Failures to flush just mean that
-                // we have some extra WAL replay to do next time the timeline starts.
-                warn!("failed to freeze and flush: {e:#}");
             }
         }
 
-        self.shutdown().await;
-    }
-
-    /// Shut down immediately, without waiting for any open layers to flush to disk.  This is a subset of
-    /// the graceful [`Timeline::flush_and_shutdown`] function.
-    pub(crate) async fn shutdown(&self) {
-        debug_assert_current_span_has_tenant_and_timeline_id();
-
-        // Signal any subscribers to our cancellation token to drop out
+        // All tasks except remote_client are sensitve to Timeline::cancel.
         tracing::debug!("Cancelling CancellationToken");
         self.cancel.cancel();
 
-        // Page request handlers might be waiting for LSN to advance: they do not respect Timeline::cancel
-        // while doing so.
-        self.last_record_lsn.shutdown();
-
-        // Shut down the layer flush task before the remote client, as one depends on the other
-        task_mgr::shutdown_tasks(
-            Some(TaskKind::LayerFlushTask),
-            Some(self.tenant_shard_id),
-            Some(self.timeline_id),
-        )
-        .await;
-
-        // Shut down remote timeline client: this gracefully moves its metadata into its Stopping state in
-        // case our caller wants to use that for a deletion
+        // Transition the remote_client into a state where it's only useful for timeline deletion.
+        // (The deletion use case is why we can't just hook up remote_client to Self::cancel).)
         if let Some(remote_client) = self.remote_client.as_ref() {
             remote_client.stop();
+            // As documented in remote_client.stop()'s doc comment, it's our responsibility
+            // to shut down the upload queue tasks.
+            // TODO: fix that, task management should be encapsulated inside remote_client.
+            task_mgr::shutdown_tasks(
+                Some(TaskKind::RemoteUploadTask),
+                Some(self.tenant_shard_id),
+                Some(self.timeline_id),
+            )
+            .await;
         }
 
-        tracing::debug!("Waiting for tasks...");
-
-        task_mgr::shutdown_tasks(None, Some(self.tenant_shard_id), Some(self.timeline_id)).await;
-
         // Finally wait until any gate-holders are complete
         self.gate.close().await;
 
+        // We're not fully confident that everything is sensitive to Self::cancel yet.
+        tracing::debug!("Waiting for tasks...");
+        let report =
+            task_mgr::shutdown_tasks(None, Some(self.tenant_shard_id), Some(self.timeline_id))
+                .await;
+        if report.planned > 0 {
+            tracing::warn!(
+                ?report,
+                "shut down stray task_mgr timeline-scoped tasks, this should not happen"
+            );
+        }
+
```

Comment:
The report generation is too costly to merge in the current form, primarily because the shutdown of the remote_timeline_client above still uses `shutdown_tasks`. But, asserting it here seems useful.

Suggestions for a compromise that we can ship before fixing remote_timeline_client (and all other users of `shutdown_tasks`? Opt-in to receive a report by passing in a parameter?